Firestore/Source/Model/FSTMutation.mm

```diff
@@ -373,37 +417,79 @@ - (nullable FSTMaybeDocument *)applyTo:(nullable FSTMaybeDocument *)maybeDoc
                                                           writeTime:(FIRTimestamp *)localWriteTime {
   NSMutableArray<FSTFieldValue *> *transformResults = [NSMutableArray array];
   for (const FieldTransform &fieldTransform : self.fieldTransforms) {
+    FSTFieldValue *previousValue = nil;
+    if ([baseDocument isMemberOfClass:[FSTDocument class]]) {
+      previousValue = [((FSTDocument *)baseDocument) fieldForPath:fieldTransform.path()];
+    }
+
+    FSTFieldValue *transformResult;
     if (fieldTransform.transformation().type() == TransformOperation::Type::ServerTimestamp) {
-      FSTFieldValue *previousValue = nil;
+      transformResult =
+          [FSTServerTimestampValue serverTimestampValueWithLocalWriteTime:localWriteTime
+                                                            previousValue:previousValue];
 
-      if ([baseDocument isMemberOfClass:[FSTDocument class]]) {
-        previousValue = [((FSTDocument *)baseDocument) fieldForPath:fieldTransform.path()];
-      }
+    } else if (fieldTransform.transformation().type() == TransformOperation::Type::ArrayUnion ||
+               fieldTransform.transformation().type() == TransformOperation::Type::ArrayRemove) {
+      transformResult =
+          [self arrayTransformResultWithPreviousValue:previousValue transform:fieldTransform];
 
-      [transformResults
-          addObject:[FSTServerTimestampValue serverTimestampValueWithLocalWriteTime:localWriteTime
-                                                                      previousValue:previousValue]];
     } else {
       FSTFail(@"Encountered unknown transform: %d type", fieldTransform.transformation().type());
     }
+
+    [transformResults addObject:transformResult];
   }
   return transformResults;
 }
 
+/**
+ * Transforms the provided `previousValue` via the provided `fieldTransform` which must be an
+ * ArrayUnion or ArrayRemove transform. Used both for local application and after server
+ * acknowledgement.
+ */
+- (FSTFieldValue *)arrayTransformResultWithPreviousValue:(FSTFieldValue *)previousValue
```

Comment:
While there's some small amount of shared code here, I think this would be clearer if there were only one thing "switching" on the transformation type (in `localTransformResultsWithBaseDocument:` with implementation methods being dedicated to one specific transformation. That will remove the extra ifs in here and remove the extra assertion, probably bringing the total line count down too.

Reply:
I went ahead with this but I'm not certain it's a net win because this code is used by both localTransformResultsWithBaseDocument: and serverTransformResultsWithBaseDocument: (a redundancy I'm not thrilled with but don't see a better alternative) and because unions and removes share a single ArrayTransform operation class...  With whitespace / comments, etc. it ended up adding 16 lines overall.  I don't feel strongly either way, but if you have second thoughts, I can revert the commit (or if you think I can do it better, I'm happy to tweak).